platform/engineering/design-docs/separate-content-build.md

```diff
@@ -0,0 +1,373 @@
+    - Runs the accessibility tests and link checker
+
+See [here](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/getting-started/workflow/deploy/)
+for more details.
+
+#### Pain points
+- Engineers encounter **confusing build errors** that are due to the mismatch in
+  outdated locally cached content and updated Liquid templates
+- Fetching content from the CMS and building HTML files when developing
+  applications **slows down the local builds** for engineers unnecessarily
+- When their applications have nothing to with the CMS
+- It **requires either SOCKS proxy access** or the `npm run fetch-drupal-cache`
+  command, which isn't obvious
+- **Application deployments and PR builds fail** when content in the CMS is
+  invalid
+  - This includes broken links and accessibility violations
+  - We have a safeguard that mostly works; if we run into a failure using the
+    latest CMS content, we try again with the cached content in S3 from the last
+    successful build
+    - This doesn't work when the GraphQL query has been updated; that
+      invalidates the cache
+    - Errors in liquid templates would potentially break the build and falling
+      back to the cache won't help
+
+### High Level Design
+The current front end build will be split up into two distinct builds:
+1. Content build
+    - Input: CMS content
+    - Output: HTML files, assets from the CMS (PDFs, images and the like)
+1. Application build
+    - Input: `vets-website` code
+    - Output: JavaScript and CSS bundles
+
+The output of both these builds will be deployed to separate buckets. The
+reverse proxies will route traffic intelligently to the appropriate bucket given
+the request.
+
+![Proposed traffic flow to S3 buckets](images/proposed-traffic-flow-to-s3-buckets.png)
+
+**Important note:** JavaScript application landing pages will live in the CMS.
+To get an application into production, the page in Drupal will need to be
+published.
+
+**Another important note:** There is no way to coordinate these two deploys to
+make an application live for the first time. The process will be to manually:
+1. Verify the application assets are live in production
+1. Publish the landing page
+1. Deploy a new content build
+
+## Specifics
+
+### Detailed Design
+
+#### Build and deploy
+- The content build will live in a separate repository
+  - **Question:** What should we call this?
+  - **Question:** Can the application build be run in Circle?
+  - **Question:** Where can we run the content build?
+- Content will be deployed to a separate S3 bucket from the application code
+- The S3 buckets will have versioned directories in them with the contents of
+  each deployment
+  - Old directories will be removed after 90 days
+    - **Question:** What makes sense? This is a totally arbitrary number
+
+#### Routing
+- The nginx will route their traffic based on some configuration file
+  - **Question:** What configuration file? How does it work?
```

Comment:
I think I'd prefer to keep configuration out of the browser. It's just a hunch, but I feel like that'd come back to bite us in some unexpected way down the road. Seems weird to be like, "Here's the version, you figure out what to go get," instead of just serving the right thing, y'know?

Reply:
FWIW I think we need to have JS/CSS in the critical rendering path referenced in the initial document response. `fetch()`ed JS has very low download priority on browsers- lower than initial doc CSS and images.